File: app/models/chunk.rb
class Chunk
  attr_reader :chapter, :chunk, :state, :input
  delegate :rule, to: :@chunk
  delegate :classification, to: :rule

  def initialize chapter, options
    @chapter = chapter
    @chunk   = chapter.assessment.chunks[options[:id]]

    @word    = options[:word]
    @error   = options[:error]
    @answer  = options[:answer]
    @grammar = options[:grammar]
    @text    = options[:text]
    @input   = options[:input]
  end

  def grade!
    return @result if defined?(@result)

    @result ||= if input? && correct.strip == @input.strip
      @state = if word? then nil else :found end
      true
    elsif !input? && word?
      @state = nil
      true
    else
      @state = if word? then :introduced else :missed end
      false
    end
  end

  def input?
    return true if @input.present? && word.blank?
    @input.present? && (@input.strip != word.strip)
  end

  def classification
    if state == :introduced
      nil
    else
      rule.classification
    end
  end

private

  def word?
    !!word
  end

  def word
    chunk[:word]
  end

  def correct
    (if chunk[:answer] then chunk[:answer] else chunk[:word] end).to_s
  end
end

File: app/assets/javascripts/backbone/models/chunk.js
var __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  __hasProp = {}.hasOwnProperty;

PG.Models.Chunk = (function(_super) {
  __extends(Chunk, _super);

  function Chunk() {
    return Chunk.__super__.constructor.apply(this, arguments);
  }

  Chunk.prototype.initialize = function(options) {
    this.word = options.word;
    this.error = options.error;
    this.answer = options.answer;
    this.grammar = options.grammar;
    return this.text = options.text;
  };

  Chunk.prototype.grade = function() {
    if (this.inputPresent() && this.correct() === this.input) {
      return true;
    } else if (!this.inputPresent() && this.word) {
      return true;
    } else {
      return false;
    }
  };

  Chunk.prototype.correct = function() {
    if (this.answer) {
      return this.answer;
    } else {
      return this.word;
    }
  };

  Chunk.prototype.inputPresent = function() {
    return _.isString(this.input) && (this.input !== this.word);
  };

  Chunk.prototype.wordDif = function(dir) {
    return diffString(this.input || this.error, this.correct());
  };

  Chunk.prototype.rule = function() {
    return chapterRules.get(this.grammar);
  };

  Chunk.prototype.css_class = function() {
    var extra;
    extra = $.trim(this.correct()) === '' || $.trim(this.correct()) === '<br>' ? ' non-editable' : '';
    return "edit-word" + extra;
  };

  Chunk.prototype.toJSON = function() {
    var json;
    return json = {
      word: this.word,
      error: this.error,
      answer: this.answer,
      grammar: this.grammar,
      text: this.text,
      input: this.input,
      id: this.attributes.id
    };
  };

  return Chunk;

})(Backbone.Model);